lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,748 @@
+			_ => Ok(InteractiveTxStateMachine {
+				context: self.context,
+				state: TheirTxComplete {}
+			})
+		}
+	}
+
+	fn send_tx_complete(self) -> InteractiveTxStateMachine<OurTxComplete> {
+		// TODO: Should we validate before transitioning states? If so, do we want to abort negotiation
+		// if our current transaction state is invalid?
+		InteractiveTxStateMachine {
+			context: self.context,
+			state: OurTxComplete {}
+		}
+	}
+}
+
+impl InteractiveTxStateMachine<OurTxComplete> {
+	fn receive_tx_complete(self) -> InteractiveTxStateMachineResult<NegotiationComplete> {
+		match self.is_current_transaction_state_able_to_complete() {
+			Err(e) => Err(InteractiveTxStateMachine { context: self.context, state: NegotiationAborted(e) }),
+			_ => Ok(InteractiveTxStateMachine {
+				context: self.context,
+				state: NegotiationComplete {}
+			})
+		}
+	}
+}
+
+impl InteractiveTxStateMachine<NegotiationComplete> {
+	fn get_psbt(&self) -> Result<Transaction, AbortReason> {
+		// Build transaction from inputs & outputs in `NegotiationContext`.
+		// TODO: Return Psbt type (?)
+		return Ok(Transaction {
+			version: self.context.base_tx.version,
+			lock_time: self.context.base_tx.lock_time,
+			input: self.context.inputs.values().map(|p| p.input.clone()).collect(),
+			output: self.context.outputs.values().cloned().collect(),
+		})
+	}
+}
+
+
+enum ChannelMode {
+	Negotiating(InteractiveTxStateMachine<Negotiating>),
+	OurTxComplete(InteractiveTxStateMachine<OurTxComplete>),
+	TheirTxComplete(InteractiveTxStateMachine<TheirTxComplete>),
+	NegotiationComplete(InteractiveTxStateMachine<NegotiationComplete>),
+	NegotiationAborted(InteractiveTxStateMachine<NegotiationAborted>),
+	Indeterminate,
+}
+
+impl Default for ChannelMode {
+	fn default() -> Self { Indeterminate }
+}
+
+pub(crate) struct InteractiveTxConstructor {
+	mode: ChannelMode,
+}
+
+// TODO: `InteractiveTxConstructor` methods should return an `Err` when the state machine itself
+// errors out. There are two scenarios where that may occur: (1) Invalid data; causing negotiation
+// to abort (2) Illegal state transition. Check spec to see if it dictates what needs to happen
+// if a node receives an unexpected message.
+impl InteractiveTxConstructor {
+	pub(crate) fn new(
+		channel_id: [u8; 32], feerate_sat_per_kw: u32, require_confirmed_inputs: bool,
+		is_initiator: bool, base_tx: Transaction, did_send_tx_signatures: bool,
+	) -> Self {
+		let initial_state_machine = InteractiveTxStateMachine::new(
+			channel_id, feerate_sat_per_kw, require_confirmed_inputs, is_initiator, base_tx,
+			did_send_tx_signatures
+		);
+		Self {
+			mode: ChannelMode::Negotiating(initial_state_machine)
+		}
+	}
+
+	pub(crate) fn abort_negotation(&mut self, reason: AbortReason) {
+		self.handle_negotiating_receive(|state_machine| state_machine.abort_negotiation(reason))
+	}
+
+	pub(crate) fn receive_tx_add_input(&mut self, serial_id: SerialId, transaction_input: &msgs::TxAddInput, confirmed: bool) {
```

Comment:
`serial_id` parameter is not used here; is it the same as `transaction_input.serial_id`? Consider removing.

Reply:
Yeah good catch, I believe that's from and old diff before introducing the full `msgs::TxAddInput` as a param.